### PR TITLE
Adding tests for previously fixed issues

### LINKED ISSUE #11835: Subquery with scalar single results are not lifted when projection has unmapped property
I've a [Guest] entity with two navigation properties, [Country] and [City], each of which has a [CountryName]/[CityName] navigation property (each Country or City has different Names based on language).

```
public class Country
{
	[Key,Column(Order = 0)]
	public Guid TenantID { get; set; }
	[Key, Column(Order = 1)]
	public Guid ID { get; set; }
	
	public virtual List<CountryName> Names { get; set; }
}
public class CountryName
{
	[Key, Column(Order = 0)]
	public Guid TenantID { get; set; }
	[Key,Column(Order = 1)]
	public Guid CountryID { get; set; }
	[Key, Column(Order = 2)]
	[StringLength(3)]
	public string LanguageID { get; set; }
	[StringLength(255)]
	public string Name { get; set; }

	public virtual Country Country { get; set; }
}

public class City
{
	[Key, Column(Order = 0)]
	public Guid TenantID { get; set; }
	[Key, Column(Order = 1)]
	public Guid CountryID { get; set; }
	[Key, Column(Order = 2)]
	public Guid ID { get; set; }
	
	public virtual List<CityName> Names { get; set; }
}
public class CityName
{
	[Key, Column(Order = 0)]
	public Guid TenantID { get; set; }
	[Key, Column(Order = 1)]
	public Guid CountryID { get; set; }
	[Key,Column(Order = 2)]
	public Guid CityID { get; set; }
	[Key, Column(Order = 3)]
	[StringLength(3)]
	public string LanguageID { get; set; }
	[StringLength(255)]
	public string Name { get; set; }

	public virtual City City { get; set; }
}

public class Guest
{
	[Key, Column(Order = 0)]
	public Guid TenantID { get; set; }
	[Key, Column(Order = 1)]
	public Guid ID { get; set; }
	public Guid? BirthCountryID { get; set; }
	public virtual Country BirthCountry { get; set; }
	public Guid? BirthCityID { get; set; }
	public virtual City BirthCity { get; set; }
}
```

[CountryName] and [CityName] have a query filter set in context OnModelCreating.

```
// Country
modelBuilder.Entity<Country>().HasKey(c => new { c.TenantID, c.ID });
modelBuilder.Entity<Country>().Property(p => p.ID).ValueGeneratedOnAdd().HasDefaultValueSql("newsequentialid()");
// CountryName
modelBuilder.Entity<CountryName>().HasKey(c => new { c.TenantID, c.CountryID, c.LanguageID });
modelBuilder.Entity<CountryName>().HasOne(ho => ho.Country).WithMany(wm => wm.Names).HasForeignKey(fk => new { fk.TenantID, fk.CountryID }).OnDelete(DeleteBehavior.Restrict);
modelBuilder.Entity<CountryName>().HasQueryFilter(qf => qf.LanguageID == LanguageID);

// City
modelBuilder.Entity<City>().HasKey(c => new { c.TenantID, c.CountryID, c.ID });
modelBuilder.Entity<City>().Property(p => p.ID).ValueGeneratedOnAdd().HasDefaultValueSql("newsequentialid()");
// CityName
modelBuilder.Entity<CityName>().HasKey(c => new { c.TenantID, c.CountryID, c.CityID, c.LanguageID });
modelBuilder.Entity<CityName>().HasOne(ho => ho.City).WithMany(wm => wm.Names).HasForeignKey(fk => new { fk.TenantID, fk.CountryID, fk.CityID }).OnDelete(DeleteBehavior.Restrict);
modelBuilder.Entity<CityName>().HasQueryFilter(qf => qf.LanguageID == LanguageID);

// Guest
modelBuilder.Entity<Guest>().HasKey(c => new { c.TenantID, c.ID });
modelBuilder.Entity<Guest>().Property(p => p.ID).ValueGeneratedOnAdd().HasDefaultValueSql("newsequentialid()");
modelBuilder.Entity<Guest>().HasOne(ho => ho.BirthCountry).WithMany().HasForeignKey(fk => new { fk.TenantID, fk.BirthCountryID }).OnDelete(DeleteBehavior.Restrict);
modelBuilder.Entity<Guest>().HasOne(ho => ho.BirthCity).WithMany().HasForeignKey(fk => new { fk.TenantID, fk.BirthCountryID, fk.BirthCityID }).OnDelete(DeleteBehavior.Restrict);
```
### MY QUERY

```
var data = _context.Guests
		.Where(w => w.TenantID == TenantID && w.Cancelled == null  && 
		(w.VAT.Contains(Filter) || w.First.Contains(Filter) || w.Last.Contains(Filter) || w.BirthDate.ToString().Contains(Filter) ||
		w.ResidenceAddress.Contains(Filter) || w.BirthCountry.Names.Any(a => a.Name.Contains(Filter)) || 
		w.BirthCity.Names.Any(a => a.Name.Contains(Filter))))
		.OrderBy(o => o.ID)
		.Skip(Start).Take(Count)
		.Select(s => new
		{
			ID = s.ID,
			Name = s.FullName,
			VAT = s.VAT,
			BirthDate = s.BirthDate.ToString("yyyy/MM/dd"),
			ResidenceAddress = s.ResidenceAddress,
			BirthCountryName = s.BirthCountry.Names[0].Name,
			BirthCityName = s.BirthCity.Names[0].Name,
			Reservations = s.ReservationsNumber,
			Bookable = bookable,
			IsHospitalized = s.IsHospitalized,
			IsPresent = s.IsPresent
		})
		.AsNoTracking()
		.ToList();
```

### GENERATED SQL
```
/* MAIN */
SELECT [w].[TenantID], [w].[ID], [w].[AddUserID], [w].[Added], [w].[BirthCityID], [w].[BirthCountryID], [w].[BirthDate], [w].[BirthState], [w].[CancelUserID], [w].[Cancelled], [w].[DeathDate], [w].[First], [w].[IsHospitalized], [w].[IsPresent], [w].[Last], [w].[ReservationsNumber] AS [Reservations], [w].[ResidenceAddress], [w].[ResidenceAddressNumber], [w].[ResidenceAddressState], [w].[ResidenceAddressZip], [w].[ResidenceCityID], [w].[ResidenceCountryID], [w].[SexID], [w].[UpdateUserID], [w].[Updated], [w].[VAT], [w.BirthCountry].[TenantID], [w.BirthCountry].[ID], [w.BirthCity].[TenantID], [w.BirthCity].[CountryID], [w.BirthCity].[ID]
FROM [Guests] AS [w]
LEFT JOIN [Cities] AS [w.BirthCity] ON (([w].[TenantID] = [w.BirthCity].[TenantID]) AND ([w].[BirthCountryID] = [w.BirthCity].[CountryID])) AND ([w].[BirthCityID] = [w.BirthCity].[ID])
LEFT JOIN [Countries] AS [w.BirthCountry] ON ([w].[TenantID] = [w.BirthCountry].[TenantID]) AND ([w].[BirthCountryID] = [w.BirthCountry].[ID])
WHERE (([w].[TenantID] = @__TenantID_0) AND [w].[Cancelled] IS NULL) AND ((((((((CHARINDEX(@__Filter_1, [w].[VAT]) > 0) OR (@__Filter_1 = N'')) OR ((CHARINDEX(@__Filter_2, [w].[First]) > 0) OR (@__Filter_2 = N''))) OR ((CHARINDEX(@__Filter_3, [w].[Last]) > 0) OR (@__Filter_3 = N''))) OR ((CHARINDEX(@__Filter_4, CONVERT(VARCHAR(100), [w].[BirthDate])) > 0) OR (@__Filter_4 = N''))) OR ((CHARINDEX(@__Filter_5, [w].[ResidenceAddress]) > 0) OR (@__Filter_5 = N''))) OR EXISTS (
    SELECT 1
    FROM [CountryNames] AS [qf]
    WHERE (([qf].[LanguageID] = @__ef_filter__LanguageID_0) AND ((CHARINDEX(@__Filter_6, [qf].[Name]) > 0) OR (@__Filter_6 = N''))) AND (([w.BirthCountry].[TenantID] = [qf].[TenantID]) AND ([w.BirthCountry].[ID] = [qf].[CountryID])))) OR EXISTS (
    SELECT 1
    FROM [CityNames] AS [qf0]
    WHERE (([qf0].[LanguageID] = @__ef_filter__LanguageID_1) AND ((CHARINDEX(@__Filter_7, [qf0].[Name]) > 0) OR (@__Filter_7 = N''))) AND ((([w.BirthCity].[TenantID] = [qf0].[TenantID]) AND ([w.BirthCity].[CountryID] = [qf0].[CountryID])) AND ([w.BirthCity].[ID] = [qf0].[CityID]))))
ORDER BY [w].[ID], [w].[TenantID], [w.BirthCountry].[TenantID], [w.BirthCountry].[ID], [w.BirthCity].[TenantID], [w.BirthCity].[CountryID], [w.BirthCity].[ID]
OFFSET @__p_8 ROWS FETCH NEXT @__p_9 ROWS ONLY

/* FOR >>>  [BirthCountryName = s.BirthCountry.Names[0].Name] */
SELECT [w.BirthCountry.Names].[TenantID], [w.BirthCountry.Names].[CountryID], [w.BirthCountry.Names].[LanguageID], [w.BirthCountry.Names].[Name], [t].[ID], [t].[TenantID], [t].[TenantID0], [t].[ID0]
FROM [CountryNames] AS [w.BirthCountry.Names]
INNER JOIN (
    SELECT [w0].[ID], [w0].[TenantID], [w.BirthCountry0].[TenantID] AS [TenantID0], [w.BirthCountry0].[ID] AS [ID0]
    FROM [Guests] AS [w0]
    LEFT JOIN [Cities] AS [w.BirthCity0] ON (([w0].[TenantID] = [w.BirthCity0].[TenantID]) AND ([w0].[BirthCountryID] = [w.BirthCity0].[CountryID])) AND ([w0].[BirthCityID] = [w.BirthCity0].[ID])
    LEFT JOIN [Countries] AS [w.BirthCountry0] ON ([w0].[TenantID] = [w.BirthCountry0].[TenantID]) AND ([w0].[BirthCountryID] = [w.BirthCountry0].[ID])
    WHERE (([w0].[TenantID] = @__TenantID_0) AND [w0].[Cancelled] IS NULL) AND ((((((((CHARINDEX(@__Filter_1, [w0].[VAT]) > 0) OR (@__Filter_1 = N'')) OR ((CHARINDEX(@__Filter_2, [w0].[First]) > 0) OR (@__Filter_2 = N''))) OR ((CHARINDEX(@__Filter_3, [w0].[Last]) > 0) OR (@__Filter_3 = N''))) OR ((CHARINDEX(@__Filter_4, CONVERT(VARCHAR(100), [w0].[BirthDate])) > 0) OR (@__Filter_4 = N''))) OR ((CHARINDEX(@__Filter_5, [w0].[ResidenceAddress]) > 0) OR (@__Filter_5 = N''))) OR EXISTS (
        SELECT 1
        FROM [CountryNames] AS [qf1]
        WHERE (([qf1].[LanguageID] = @__ef_filter__LanguageID_0) AND ((CHARINDEX(@__Filter_6, [qf1].[Name]) > 0) OR (@__Filter_6 = N''))) AND (([w.BirthCountry0].[TenantID] = [qf1].[TenantID]) AND ([w.BirthCountry0].[ID] = [qf1].[CountryID])))) OR EXISTS (
        SELECT 1
        FROM [CityNames] AS [qf2]
        WHERE (([qf2].[LanguageID] = @__ef_filter__LanguageID_1) AND ((CHARINDEX(@__Filter_7, [qf2].[Name]) > 0) OR (@__Filter_7 = N''))) AND ((([w.BirthCity0].[TenantID] = [qf2].[TenantID]) AND ([w.BirthCity0].[CountryID] = [qf2].[CountryID])) AND ([w.BirthCity0].[ID] = [qf2].[CityID]))))
    ORDER BY [w0].[ID], [w0].[TenantID], [w.BirthCountry0].[TenantID], [w.BirthCountry0].[ID]
    OFFSET @__p_8 ROWS FETCH NEXT @__p_9 ROWS ONLY
) AS [t] ON ([w.BirthCountry.Names].[TenantID] = [t].[TenantID0]) AND ([w.BirthCountry.Names].[CountryID] = [t].[ID0])
WHERE [w.BirthCountry.Names].[LanguageID] = @__ef_filter__LanguageID_2
ORDER BY [t].[ID], [t].[TenantID], [t].[TenantID0], [t].[ID0]

/* FOR >>> [BirthCityName = s.BirthCity.Names[0].Name]*/
SELECT [w.BirthCity.Names].[TenantID], [w.BirthCity.Names].[CountryID], [w.BirthCity.Names].[CityID], [w.BirthCity.Names].[LanguageID], [w.BirthCity.Names].[Name], [t0].[ID], [t0].[TenantID], [t0].[TenantID0], [t0].[CountryID], [t0].[ID0]
FROM [CityNames] AS [w.BirthCity.Names]
INNER JOIN (
    SELECT [w1].[ID], [w1].[TenantID], [w.BirthCity1].[TenantID] AS [TenantID0], [w.BirthCity1].[CountryID], [w.BirthCity1].[ID] AS [ID0]
    FROM [Guests] AS [w1]
    LEFT JOIN [Cities] AS [w.BirthCity1] ON (([w1].[TenantID] = [w.BirthCity1].[TenantID]) AND ([w1].[BirthCountryID] = [w.BirthCity1].[CountryID])) AND ([w1].[BirthCityID] = [w.BirthCity1].[ID])
    LEFT JOIN [Countries] AS [w.BirthCountry1] ON ([w1].[TenantID] = [w.BirthCountry1].[TenantID]) AND ([w1].[BirthCountryID] = [w.BirthCountry1].[ID])
    WHERE (([w1].[TenantID] = @__TenantID_0) AND [w1].[Cancelled] IS NULL) AND ((((((((CHARINDEX(@__Filter_1, [w1].[VAT]) > 0) OR (@__Filter_1 = N'')) OR ((CHARINDEX(@__Filter_2, [w1].[First]) > 0) OR (@__Filter_2 = N''))) OR ((CHARINDEX(@__Filter_3, [w1].[Last]) > 0) OR (@__Filter_3 = N''))) OR ((CHARINDEX(@__Filter_4, CONVERT(VARCHAR(100), [w1].[BirthDate])) > 0) OR (@__Filter_4 = N''))) OR ((CHARINDEX(@__Filter_5, [w1].[ResidenceAddress]) > 0) OR (@__Filter_5 = N''))) OR EXISTS (
        SELECT 1
        FROM [CountryNames] AS [qf3]
        WHERE (([qf3].[LanguageID] = @__ef_filter__LanguageID_0) AND ((CHARINDEX(@__Filter_6, [qf3].[Name]) > 0) OR (@__Filter_6 = N''))) AND (([w.BirthCountry1].[TenantID] = [qf3].[TenantID]) AND ([w.BirthCountry1].[ID] = [qf3].[CountryID])))) OR EXISTS (
        SELECT 1
        FROM [CityNames] AS [qf4]
        WHERE (([qf4].[LanguageID] = @__ef_filter__LanguageID_1) AND ((CHARINDEX(@__Filter_7, [qf4].[Name]) > 0) OR (@__Filter_7 = N''))) AND ((([w.BirthCity1].[TenantID] = [qf4].[TenantID]) AND ([w.BirthCity1].[CountryID] = [qf4].[CountryID])) AND ([w.BirthCity1].[ID] = [qf4].[CityID]))))
    ORDER BY [w1].[ID], [w1].[TenantID], [w.BirthCity1].[TenantID], [w.BirthCity1].[CountryID], [w.BirthCity1].[ID]
    OFFSET @__p_8 ROWS FETCH NEXT @__p_9 ROWS ONLY
) AS [t0] ON (([w.BirthCity.Names].[TenantID] = [t0].[TenantID0]) AND ([w.BirthCity.Names].[CountryID] = [t0].[CountryID])) AND ([w.BirthCity.Names].[CityID] = [t0].[ID0])
WHERE [w.BirthCity.Names].[LanguageID] = @__ef_filter__LanguageID_3
ORDER BY [t0].[ID], [t0].[TenantID], [t0].[TenantID0], [t0].[CountryID], [t0].[ID0]
```

### QUESTION
Actually 3 query are generated, 3 resultset are returned from SQL Server (Azure) and then client side merged.

Maybe this query could be translated to a single SQL query ?
Or at least select only [Name] field from nested navigation properties ?
(specially on mobile devices any byte saved is a great achievement)

I tried 
`BirthCityName = s.BirthCity.Names.SingleOrDefault().Name`
or
`BirthCityName = s.BirthCity.Names.FirstOrDefault().Name`
instead of
`BirthCityName = s.BirthCity.Names[0].Name`
but both aren't translated and are resolved client side, generating [Count]  (Take parameter) * 2 (Country and City) more SQL queries, two for each guest retrieved (taking 25 records EF Core generate 51 SQL queries)

### Further technical details
EF Core version: 2.1.0-preview2-final
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system:  Windows 10 64 bit
IDE:  Visual Studio 2017 15.7.0 Preview 5

### LINKED ISSUE #12274: Initializing a non-entity ref type in select creates a single instance for every record
Consider the following entity type and its DTO type.
```csharp
// Entity
public class Dog
{
    public int Id { get; set; }
    public string Name { get; set; }
}

//DTO
 public class DogDTO
 {
     public int ID { get; set; }
     public string  Name { get; set; }
     public IDCard IDCard { get; set; }
 }

 public class IDCard
 {
     public int ID { get; set; }
     public string Number { get; set; }
 }
```
Now query all the entities and project them to the DTO type.

```csharp
 var dogDTOList = ctx.Dogs.Select(
   d => 
     new DogDTO { ID = d.Id
    ,Name = d.Name
    ,IDCard = new IDCard() })
.ToList();
```
We would expect that a new IDCard instance will be created for each DogDTO - as this is what would happen in case of LINQ-to-Objects.
However, the constructor of IDCard is called only once and the IDCard instance will be the same for all DogDTOs.

There is a full repro repo [here](https://github.com/kszicsillag/efcore-selectinit).

### Further technical details
EF Core version: 2.1.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows Server 2016
IDE: Visual Studio 2017 15.7

### LINKED ISSUE #12379: Query: optimize correlated subquery cases where inner collection references query source two levels up
Currently, query like:

```cs
from o in ctx.Gears.OfType<Officer>()
                    select new
                    {
                        o.FullName,
                        Collection = (from r in o.Reports
                                      select new
                                      {
                                          ReportName = r.FullName,
                                          OfficerName = o.FullName
                                      }).ToList()
```

doesn't take advantage of correlated subquery optimization, because innermost subquery refences o.FullName. If the `OfficerName` line is removed from the innermost projection, we generate the following:

```
SELECT [t].[Nickname], [t].[SquadId], [o.Reports].[FullName] AS [ReportName], [o.Reports].[LeaderNickname], [o.Reports].[LeaderSquadId]
                        FROM [Gears] AS [o.Reports]
                        INNER JOIN (
                            SELECT [o0].[Nickname], [o0].[SquadId]
                            FROM [Gears] AS [o0]
                            WHERE [o0].[Discriminator] = N'Officer'
                        ) AS [t] ON ([o.Reports].[LeaderNickname] = [t].[Nickname]) AND ([o.Reports].[LeaderSquadId] = [t].[SquadId])
                        WHERE [o.Reports].[Discriminator] IN (N'Officer', N'Gear')
                        ORDER BY [t].[Nickname], [t].[SquadId]
```

In theory, we could extract o.FullName from the INNER JOIN part of the query - we just need to be smart about remapping it correctly.

### LINKED ISSUE #12597: Query: Accessing Length property on a subquery projecting string with FirstOrDefault could be pushed into the subquery
query:

```cs
ctx.Customers.Select(c => c.Orders.OrderBy(o => o.OrderID).Select(o => o.CustomerID).FirstOrDefault().Length)
```

produces:

```
(QueryContext queryContext) => IEnumerable<int> _InterceptExceptions(
|__ source: IEnumerable<int> _ShapedQuery(
|   |__ queryContext: queryContext, 
|   |__ shaperCommandContext: SelectExpression: 
|   |       SELECT (
|   |           SELECT TOP(1) [o].[CustomerID]
|   |           FROM [Orders] AS [o]
|   |           WHERE [c].[CustomerID] = [o].[CustomerID]
|   |           ORDER BY [o].[OrderID]
|   |       )
|   |       FROM [Customers] AS [c], 
|   |__ shaper: (QueryContext queryContext | ValueBuffer c) => ((string)string TryReadValue(c, 0, null)).Length), 
|__ contextType: TestModels.Northwind.NorthwindRelationalContext, 
|__ logger: DiagnosticsLogger<Query>, 
|__ queryContext: Unhandled parameter: queryContext)
```

It should be safe to optimize to:

```cs
ctx.Customers.Select(c => c.Orders.OrderBy(o => o.OrderID).Select(o => o.CustomerID.Length).FirstOrDefault())
```

which currently produces:

```
(QueryContext queryContext) => IEnumerable<int> _InterceptExceptions(
|__ source: IEnumerable<int> _ShapedQuery(
|   |__ queryContext: queryContext, 
|   |__ shaperCommandContext: SelectExpression: 
|   |       SELECT COALESCE((
|   |           SELECT TOP(1) CAST(LEN([o].[CustomerID]) AS int)
|   |           FROM [Orders] AS [o]
|   |           WHERE [c].[CustomerID] = [o].[CustomerID]
|   |           ORDER BY [o].[OrderID]
|   |       ), 0)
|   |       FROM [Customers] AS [c], 
|   |__ shaper: (QueryContext queryContext | ValueBuffer c) => (int)Nullable<int> TryReadValue(c, 0, null)), 
|__ contextType: TestModels.Northwind.NorthwindRelationalContext, 
|__ logger: DiagnosticsLogger<Query>, 
|__ queryContext: Unhandled parameter: queryContext)
```

### LINKED ISSUE #12717: EF generates different SQL queries for expressions with type conversion using () operator vs 'as' operator
The generated SQL queries are different for an expression with type conversion using () operator vs. an expression with 'as' operator.

Please refer to the following code snippets and relevent log outputs.

**Code**
```
context.Set<Animal>().Where(x => x is Dog && ((Dog)x).Name == "Woof").ToList();
```
**Output**
```
 SELECT [x].[Id], [x].[AnimalType], [x].[Name]
 FROM [Animal] AS [x]
 WHERE [x].[AnimalType] IN (3, 2, 1) AND (([x].[AnimalType] = 3) AND ([x].[Name] = N'Woof'))
```

**Code**
```
context.Set<Animal>().Where(x => x is Dog && (x as Dog).Name == "Woof").ToList();
```
**Output**
```
SELECT [x].[Id], [x].[AnimalType], [x].[Name]
FROM [Animal] AS [x]
WHERE [x].[AnimalType] IN (3, 2, 1) AND ([x].[AnimalType] = 3)
```

### Steps to reproduce
Repro included at https://github.com/avinash-phaniraj-readify/EFCoreTestBed/tree/TypeConversionWithAsOperatorError

### Further technical details
EF Core version: 2.1
Database Provider: EntityFrameworkCore.SqlServerCompact35
Operating system: Windows 10.0.17134
IDE: Visual Studio 2017 15.7.4

### LINKED ISSUE #17775: Average_on_nav_subquery_in_projection is flaky
>Microsoft.Data.SqlClient.SqlException : The incoming tabular data stream (TDS) protocol stream is incorrect. The MARS TDS header contained errors.
   at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__164_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken) in /_/src/EFCore.Relational/Storage/RelationalCommand.cs:line 543
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken) in /_/src/EFCore.Relational/Storage/RelationalCommand.cs:line 543
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken) in /_/src/EFCore.Relational/Storage/RelationalCommand.cs:line 543
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.AsyncQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync() in /_/src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.AsyncQueryingEnumerable.cs:line 99
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken) in /_/src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs:line 2019
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken) in /_/src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs:line 2019
   at Microsoft.EntityFrameworkCore.Query.AsyncSimpleQueryTestBase`1.Average_on_nav_subquery_in_projection() in /_/test/EFCore.Specification.Tests/Query/AsyncSimpleQueryTestBase.cs:line 143
--- End of stack trace from previous location where exception was thrown ---

### LINKED ISSUE #18962: SkipWhile: Processing of the LINQ expression failed
### Steps to reproduce

``` c#
using Microsoft.EntityFrameworkCore;
using System;
using System.Linq;

namespace ConsoleApp3
{
    public sealed class Record
    {
        public int Id { get; set; }
        public DateTime Time { get; set; }
    }

    public sealed class ApplicationDbContext : DbContext
    {
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            base.OnConfiguring(optionsBuilder);
            optionsBuilder.UseSqlServer("Server=(localdb)\\mssqllocaldb;Database=aspnet-WebApplication-6E483A89-BEE6-4A76-96CC-CEB276E5E112;Trusted_Connection=True;MultipleActiveResultSets=true");
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);
        }

        public DbSet<Record> Records { get; set; }
    }

    public static class Program
    {
        public static void Main()
        {
            using (var context = new ApplicationDbContext())
            {
                context.Database.EnsureDeleted();
                context.Database.EnsureCreated();

                context.Records.Add(new Record { Time = new DateTime(2019, 11, 16) });
                context.Records.Add(new Record { Time = new DateTime(2019, 11, 17) });
                context.Records.Add(new Record { Time = new DateTime(2019, 11, 18) });
                context.SaveChanges();
            }

            using (var context = new ApplicationDbContext())
            {
                var record = context.Records.Find(2);

                var previousRecord = context.Records
                    .OrderByDescending(r => r.Time)
                    .ThenByDescending(r => r.Id)
                    .SkipWhile(r => r.Id != record.Id)
                    .Skip(1)
                    .FirstOrDefault();
            }
        }
    }
}
```

After running this, I get this exception:

> System.InvalidOperationException: 'Processing of the LINQ expression 'SkipWhile<Record>(
    source: ThenByDescending<Record, int>(
        source: OrderByDescending<Record, DateTime>(
            source: DbSet<Record>, 
            keySelector: (r) => r.Time), 
        keySelector: (r) => r.Id), 
    predicate: (r) => r.Id != (Unhandled parameter: __record_Id_0))' by 'NavigationExpandingExpressionVisitor' failed. This may indicate either a bug or a limitation in EF Core. See https://go.microsoft.com/fwlink/?linkid=2101433 for more detailed information.'

I assume this is because `SkipWhile` isn't actually implemented (looking at #14104), but the error message is really cryptic. It's an `InvalidOperationException` simply saying that processing the expression failed and mentions that it might be a bug in EF Core.

Running the same code under EF Core 2.2, I get:

> System.NotSupportedException: 'Could not parse expression 'value(Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1[ConsoleApp3.Record]).OrderByDescending(r => r.Time).ThenByDescending(r => r.Id).SkipWhile(r => (r.Id != __record_Id_0))': This overload of the method 'System.Linq.Queryable.SkipWhile' is currently not supported.'

which is much clearer - it tells me that `SkipWhile` isn't supported. This is at the very least a regression in the readability of this error message.

### Further technical details

EF Core version: 3.0.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.0
Operating system: Windows 10 Version 1903
IDE: Visual Studio 16.4.0 Preview 5.0

### PR DESCRIPTION
**Pull request by maumar**
Resolves #12274
Resolves #11835
Resolves #12597
Resolves #12379
Resolves #12717
Resolves #17775
Resolves #18962

### CODE REVIEW

File: test/EFCore.InMemory.FunctionalTests/Query/NorthwindAggregateOperatorsQueryInMemoryTest.cs [hunk 1]
```diff
@@ -54,6 +54,14 @@ public override void Min_no_data_subquery()
                     () => context.Customers.Select(c => c.Orders.Where(o => o.OrderID == -1).Min(o => o.OrderID)).ToList()).Message);
         }
 
+        public override async Task Average_on_nav_subquery_in_projection(bool isAsync)
+        {
+            Assert.Equal(
+                "Sequence contains no elements",
```
Comment (smitpatel) [inline test/EFCore.InMemory.FunctionalTests/Query/NorthwindAggregateOperatorsQueryInMemoryTest.cs:60]: We have this message in CoreStrings also.

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs [hunk 2]
```diff
@@ -10,6 +10,7 @@
 using System.Linq;
 using System.Linq.Expressions;
 using System.Reflection;
+using System.Security.Policy;
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:13]: Interesting using

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs [hunk 3]
```diff
@@ -23,6 +24,7 @@
 using Microsoft.Extensions.Caching.Memory;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Logging;
+using ModelSnapshot22;
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:27]: Interesting using

**Comment by smitpatel** [comment]
#17775 was flaky and underlying issue may not be gone. Make sure to run all the tests in SqlServer functional tests multiple times to verify that it is not flaky.

**Review by smitpatel — APPROVED** [review]

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs (hunk 2 above)
Comment (maumar) [reply test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:13]: VS "feature"

**Comment by maumar** [comment]
@msftbot require 1 approval

**Pull request merged**